### ISSUE TITLE
Does GRPCLogReportServiceClient support TLS?

### ISSUE THREAD

**Issue by BFergerson** [apache/skywalking#7040]
Please answer these questions before submitting your issue.

- Why do you submit this issue?
- [x] Question or discussion

___
### Question

I'm able to send traces via TLS but when I try to send logs, I get the following exception:
```
ERROR 2021-05-30 22:21:59:387 grpc-default-executor-1 GRPCLogReportServiceClient : Try to send 1 log data to collector, with unexpected exception.
org.apache.skywalking.apm.dependencies.io.grpc.StatusRuntimeException: UNAVAILABLE: Network closed for unknown reason
        at org.apache.skywalking.apm.dependencies.io.grpc.Status.asRuntimeException(Status.java:533)
        at org.apache.skywalking.apm.dependencies.io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:478)
        at org.apache.skywalking.apm.dependencies.io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:413)
        at org.apache.skywalking.apm.dependencies.io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
        at org.apache.skywalking.apm.dependencies.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
        at org.apache.skywalking.apm.dependencies.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:721)
        at org.apache.skywalking.apm.dependencies.io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at org.apache.skywalking.apm.dependencies.io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
```

I've tried to debug this issue and ran into:
https://github.com/apache/skywalking/blob/bd38ec492e013fec7ffc1359c9c8a7e9e1d52389/apm-sniffer/apm-toolkit-activation/apm-toolkit-logging-common/src/main/java/org/apache/skywalking/apm/toolkit/logging/common/log/GRPCLogReportServiceClient.java#L71

Is that the reason TLS isn't working and if so, is there any reason why logs shouldn't also be sent through TLS?

**Comment by wu-sheng** [comment]
Same as #6664. We plan to merge the channel, once Satellite support all metrics/logs/traces/event/management channels.

**Comment by wu-sheng** [comment]
If you want to add this, it is fine, but we have to expose more configs.

**Comment by BFergerson** [comment]
It looks like I may have to wait on this one as I have no experience with Go and from reading the ticket you linked, there needs to be work on the Sidecar project to implement this functionality. Would you accept a PR that fixes this issue via additional configuration params for those without Sidecar and if so, how would you fix `incompatible types: org.apache.skywalking.apm.dependencies.io.grpc.Channel cannot be converted to io.grpc.Channel`?

**Comment by wu-sheng** [comment]
Adding config is fine. What do you mean converting error? It seems a classloader issue.

**Comment by BFergerson** [comment]
I'm not completely sure how I should be implementing this but I was just testing adding `TLSChannelBuilder` and I get an error like above.

For example, changing this:
```java
channel = ManagedChannelBuilder
            .forAddress(
                ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_HOST,
                ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_PORT
            )
            .usePlaintext()
            .build();
```

To this:
```java
channel = new TLSChannelBuilder().build(NettyChannelBuilder
            .forAddress(
                ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_HOST,
                ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_PORT
            ))
            .build();
```

Causes this:
```
incompatible types: io.grpc.netty.NettyChannelBuilder cannot be converted to org.apache.skywalking.apm.dependencies.io.grpc.netty.NettyChannelBuilder
```

I've tried playing around with the shade settings for `apm-toolkit-logging-common` but nothing seems to fix it. I'd appreciate a little direction.

**Comment by wu-sheng** [comment]
Have you checked what does compiled codes look like? Such as whether `io.grpc.netty.NettyChannelBuilder` really gets shared in `GRPCLogReportServiceClient`.

**Comment by BFergerson** [comment]
That's the thing though, I can't get it to compile using `TLSChannelBuilder`. If I copy the code from that class into `GRPCLogReportServiceClient` I can confirm that it works but I'm having trouble getting it to compile using `TLSChannelBuilder` directly. I'm unsure how to modify the shade settings to correct this.

**Comment by wu-sheng** [comment]
My point is, you need to find out, which class(SkyWalking's) compiling has issues with `shade`. From the error you got, `io.grpc.netty.NettyChannelBuilder` is not being named. `io.grpc` is already on the shade list, so, where goes wrong? We need to focus on this.

**Comment by zhyyu** [comment]
> That's the thing though, I can't get it to compile using `TLSChannelBuilder`. If I copy the code from that class into `GRPCLogReportServiceClient`, I can confirm that the TLS connection works but I'm having trouble getting it to compile using `TLSChannelBuilder` directly. I'm unsure how to modify the shade settings to correct this.


You can use java reflection to avoid the shade compile problem, like followed, I try to fix the log authentication [issue](https://github.com/apache/skywalking/issues/6664) on my local machine.

from
```java
GRPCLogReportServiceClient

channel = ManagedChannelBuilder
            .forAddress(
                ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_HOST,
                ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_PORT
            )
            .usePlaintext()
            .build();
        asyncStub = LogReportServiceGrpc.newStub(channel)
                                        .withMaxOutboundMessageSize(
                                            ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.MAX_MESSAGE_SIZE);
```

to 

```java
GRPCLogReportServiceClient

channel = ManagedChannelBuilder
            .forAddress(
                ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_HOST,
                ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.SERVER_PORT
            )
            .usePlaintext()
            .build();

        // use reflection to avoid shade compile error
        AuthenticationDecorator authenticationDecorator = new AuthenticationDecorator();
        Method method = AuthenticationDecorator.class.getMethod("build", Channel.class);
        Channel decoratedChannel = (Channel) method.invoke(authenticationDecorator, channel);

        asyncStub = LogReportServiceGrpc.newStub(decoratedChannel)
                                        .withMaxOutboundMessageSize(
                                            ToolkitConfig.Plugin.Toolkit.Log.GRPC.Reporter.MAX_MESSAGE_SIZE);
```

**Comment by wu-sheng** [comment]
Same as #6664

**Comment by wu-sheng** [comment]
@dmsolr What is status now?

**Comment by dmsolr** [comment]
As for TLS over gRPC channel, on OAP server support TLS over OAP server nodes communication and between OAP server and Probes. In other word, the `core` module and `receiver-sharing-server` module.

**Comment by dmsolr** [comment]
After #7565 merged, SkyWalking supports mTLS mode between Probe and OAP Servers. 
And Java agent([skywalking-java#15](https://github.com/apache/skywalking-java/pull/15)) and Satellite have supported mTLS.

For LogReporter in the toolkit, I will submit a new PR to combine the reporter channel with Agent main channel.

**Comment by wu-sheng** [comment]
@mrproliu We are going to release Satellite 0.2.0 in next 2 weeks, right?

**Comment by mrproliu** [comment]
> @mrproliu We are going to release Satellite 0.2.0 in next 2 weeks, right?

Yes. I'll perfect some documents before release.

**Comment by wu-sheng** [comment]
Supported in agent side, as log report channel has been merged into core channel.
